search-provider/yelp-search-provider.c

```diff
@@ -0,0 +1,603 @@
+        g_dbus_method_invocation_return_error_literal (invocation,
+                                                       G_IO_ERROR,
+                                                       G_IO_ERROR_FAILED,
+                                                       "failed to get connection to the session bus");
+        return;
+    }
+
+    g_variant_builder_init (&param_builder, G_VARIANT_TYPE ("av"));
+    g_variant_builder_add (&param_builder, "v", g_variant_new_string (string_parameter));
+
+    reply = g_dbus_connection_call_sync (session_bus,
+                                         "org.gnome.Yelp",
+                                         "/org/gnome/Yelp",
+                                         "org.freedesktop.Application",
+                                         "ActivateAction",
+                                         g_variant_new ("(sava{sv})",
+                                                        action,
+                                                        &param_builder,
+                                                        NULL),
+                                         NULL,
+                                         G_DBUS_CALL_FLAGS_NONE,
+                                         -1,
+                                         NULL,
+                                         &error);
+
+    if (reply)
+        g_variant_unref (reply);
+    if (error) {
+        g_dbus_method_invocation_return_gerror (invocation, error);
+        g_error_free (error);
+        return;
+    }
+    g_dbus_method_invocation_return_value (invocation, NULL);
+    return;
+}
+
+static gboolean 
+handle_launch_search (YelpShellSearchProvider2  *skeleton,
+                      GDBusMethodInvocation     *invocation,
+                      gchar                    **terms,
+                      guint32                    timestamp,
+                      gpointer                   user_data)
+{
+    YelpSearchProviderApp *app = YELP_SEARCH_PROVIDER_APP (user_data);
+    gchar *joined_terms = g_strjoinv (" ", terms);
+
+    yelp_launch (joined_terms, app, invocation, "show-search");
+
+    g_free (joined_terms);
+
+    return TRUE;
+}
+
+static gboolean
+handle_activate_result (YelpShellSearchProvider2  *skeleton,
+                        GDBusMethodInvocation     *invocation,
+                        gchar                     *result,
+                        gchar                    **terms,
+                        guint32                    timestamp,
+                        gpointer                   user_data)
+{
+    YelpSearchProviderApp *app = YELP_SEARCH_PROVIDER_APP (user_data);
+
+    yelp_launch (result, app, invocation, "show-page");
+
+    return TRUE;
+}
+
+static void
+search_provider_app_dispose (GObject *obj)
+{
+    YelpSearchProviderApp *self = YELP_SEARCH_PROVIDER_APP (obj);
+
+    g_clear_object (&self->skeleton);
+    g_clear_pointer (&self->page_data_hash_map, g_hash_table_unref);
```

Comment:
Yeah, instead of the connection ID, you can also simply store the URI pointer in the private struct and use `g_signal_handlers_disconnect_by_func()` instead when the URI is not `NULL`.

`g_signal_connect_object()` also would probably work though, so feel free to use that instead.